apps/confidential/confidential.go

```diff
@@ -0,0 +1,364 @@
+				if block == nil {
+					return nil, nil, fmt.Errorf("encounter encrypted PEM block that did not decode")
+				}
+			} else {
+				return nil, nil, fmt.Errorf("encountered encrypted PEM block, but password was not passed")
+			}
+		}
+
+		switch block.Type {
+		case "CERTIFICATE":
+			cert, err := x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				return nil, nil, fmt.Errorf("block labelled 'CERTIFICATE' could not be pared by x509: %w", err)
+			}
+			certs = append(certs, cert)
+		case "PRIVATE KEY":
+			if priv != nil {
+				return nil, nil, fmt.Errorf("found multiple blocks labelled 'PRIVATE KEY'")
+			}
+
+			var err error
+			priv, err = parsePrivateKey(block.Bytes)
+			if err != nil {
+				return nil, nil, fmt.Errorf("could not decode private key: %w", err)
+			}
+		}
+		pemData = rest
+	}
+
+	if len(certs) == 0 {
+		return nil, nil, fmt.Errorf("no certificates found")
+	}
+
+	if priv == nil {
+		return nil, nil, fmt.Errorf("no private key found")
+	}
+
+	return certs, priv, nil
+}
+
+// parsePrivateKey is based on https://gist.github.com/ukautz/cd118e298bbd8f0a88fc . I don't *think*
+// we need to do the extra decoding in the example, but *maybe*?
+func parsePrivateKey(der []byte) (crypto.PrivateKey, error) {
+	key, err := x509.ParsePKCS8PrivateKey(der)
+	if err != nil {
+		return nil, fmt.Errorf("problems decoding private key using PKCS8: %w", err)
+	}
+	return key, nil
+}
+
+// Credential represents the credential used in confidential client flows.
+type Credential struct {
+	secret string
+
+	cert *x509.Certificate
+	key  crypto.PrivateKey
+}
+
+// toMSALBASE returns the accesstokens.Credential that is used internally. The current structure of the
+// code requires that client.go, requests.go and confidential.go share a credential type without
+// having import recursion. That requires the type used between is in a shared package. Therefore
+// we have this.
+// TODO(jdoak): change method name.
+func (c Credential) toMSALBASE() *accesstokens.Credential {
+	return &accesstokens.Credential{Secret: c.secret, Cert: c.cert, Key: c.key}
+}
+
+// NewCredFromSecret creates a Credential from a secret.
+func NewCredFromSecret(secret string) (Credential, error) {
+	if secret == "" {
+		return Credential{}, errors.New("secret can't be empty string")
+	}
+	return Credential{secret: secret}, nil
+}
+
+// NewCredFromCert creates a Credential from an x509.Certificate and a PKCS8 DER encoded private key.
+// CertFromPEM() can be used to get these values from a PEM file storing a PKCS8 private key.
+func NewCredFromCert(cert *x509.Certificate, key crypto.PrivateKey) Credential {
+	return Credential{cert: cert, key: key}
+}
+
+// Client is a representation of authentication client for confidential applications as defined in the
+// package doc. A new Client should be created PER SERVICE USER.
+// For more information, visit https://docs.microsoft.com/azure/active-directory/develop/msal-client-applications
+type Client struct {
```

Comment:
How do I specify a custom HTTP client?  Customers will need this so they can inject things like retry logic and logging.

Reply:
we discussed this before. we had differences in opinion on this. We should ensure we can inject the  HttpClient, it's what we promise for MSAL's

Reply:
One of the things discussed at the meeting was an elimination of custom http logic until a customer comes with a use case that can't be implemented in other ways (such as passing a Retrier via a context object).

I've eliminated that ability for the time being.  However it should be trivial to add in the future as a constructor option.

The previous interface I saw was not up to snuff (the one that had like Get() and Post()).  Any interface should mimic *http.Client methods exactly.  Aka:
``` go
// HTTP provides an HTTP client that will be used for all HTTP calls. This interface is not covered by any
// compatibility promise. Additions to the stdlib http.Client will be added here and may break compatibility without
// a semantic version change.
type HTTP interface {
    CloseIdleConnections()
    Do(req *Request) (*http.Response, error)
    Get(url string) (*http.Response, error)
    Head(url string) (*http.Response, error)
    Post(url, contentType string, body io.Reader) (*http.Response,error)
    PostForm(url string, data url.Values) (*http.Response,error)
}
```

Reply:
Tracking this here: #119